generator/internal/discovery_resource.cc

```diff
@@ -26,13 +27,117 @@ namespace google {
 namespace cloud {
 namespace generator_internal {
 
+DiscoveryResource::DiscoveryResource() : json_("") {}
+
 DiscoveryResource::DiscoveryResource(std::string name, std::string default_host,
                                      std::string base_path, nlohmann::json json)
     : name_(std::move(name)),
       default_host_(std::move(default_host)),
       base_path_(std::move(base_path)),
       json_(std::move(json)) {}
 
+StatusOr<std::string> DiscoveryResource::FormatRpcOptions(
+    nlohmann::json const& method_json,
+    DiscoveryTypeVertex const& request_type) const {
+  std::vector<std::string> rpc_options;
+  std::string verb = absl::AsciiStrToLower(method_json.value("httpMethod", ""));
+  std::string path = method_json.value("path", "");
+
+  if (verb.empty() || path.empty()) {
+    return internal::InvalidArgumentError(
+        "Method does not define httpMethod and/or path.",
+        GCP_ERROR_INFO().WithMetadata("json", method_json.dump()));
+  }
+
+  std::string http_option = absl::StrFormat(
+      "    option (google.api.http) = {\n"
+      "      %s: \"%s\"\n",
+      verb, absl::StrCat(absl::StripSuffix(base_path_, "/"), "/", path));
+  std::string request_resource_field_name;
+  if (verb == "post" || verb == "patch" || verb == "put") {
+    std::string http_body =
+        request_type.json().value("request_resource_field_name", "*");
+    if (http_body != "*") request_resource_field_name = http_body;
+    absl::StrAppend(&http_option,
+                    absl::StrFormat("      body: \"%s\"\n", http_body));
+  }
+  rpc_options.push_back(absl::StrCat(http_option, "    };"));
+
+  std::string operation_scope;
+  std::vector<std::string> params =
+      method_json.value("parameterOrder", std::vector<std::string>{});
+  if (!params.empty()) {
```

Comment:
I don't see any tests for the case where `params.empty() == true`.

As written, it could have an `option (google.cloud.operation_service) = "";`. Is that right?

If not, consider just: `if (params.empty()) return rpc_options.front();`